src/assets/scripts/client/aircraft/FlightManagementSystem/RouteModel.js

```diff
@@ -0,0 +1,902 @@
+         * exist, at which point they will simply hold their last assigned heading and altitude.
+         *
+         * @for RouteModel
+         * @property _previousLegCollection
+         * @type {array<WaypointModel>}
+         * @private
+         */
+        this._previousLegCollection = [];
+
+        this.init(routeString);
+    }
+
+    /**
+     * Return the current `LegModel`
+     *
+     * @for RouteModel
+     * @property currentLeg
+     * @type {LegModel}
+     */
+    get currentLeg() {
+        if (this._legCollection.length < 1) {
+            throw new TypeError('Expected the route to contain at least one leg');
+        }
+
+        return this._legCollection[0];
+    }
+
+    /**
+     * Return the current `WaypointModel`
+     *
+     * @for RouteModel
+     * @property currentWaypoint
+     * @type {WaypointModel}
+     */
+    get currentWaypoint() {
+        return this.currentLeg.currentWaypoint;
+    }
+
+    /**
+     * Return the next `LegModel`, if it exists
+     *
+     * @for RouteModel
+     * @property nextLeg
+     * @type {LegModel}
+     */
+    get nextLeg() {
+        if (!this.hasNextLeg()) {
+            return null;
+        }
+
+        return this._legCollection[1];
+    }
+
+    /**
+     * Return the next `WaypointModel`, from current or future leg
+     *
+     * @for RouteModel
+     * @property nextWaypoint
+     * @type {WaypointModel}
+     */
+    get nextWaypoint() {
+        if (!this.hasNextWaypoint()) {
+            return null;
+        }
+
+        if (this.currentLeg.hasNextWaypoint()) {
+            return this.currentLeg.nextWaypoint;
+        }
+
+        return this.nextLeg.currentWaypoint;
+    }
+
+    /**
+     * Return an array of all waypoints in all legs of the route
+     *
+     * @for RouteModel
+     * @property waypoints
+     * @type {array<WaypointModel>}
+     */
+    get waypoints() {
+        return _reduce(this._legCollection, (waypointList, legModel) => {
+            return waypointList.concat(legModel.waypoints);
+        }, []);
+    }
+
+    // ------------------------------ LIFECYCLE ------------------------------
+
+    /**
+     * Initialize class properties
+     *
+     * @for RouteModel
+     * @method init
+     * @param routeString {string}
+     * @chainable
+     */
+    init(routeString) {
+        this._legCollection = this._generateLegsFromRouteString(routeString);
+
+        return this;
+    }
+
+    /**
+     * Reset class properties
+     *
+     * @for RouteModel
+     * @method reset
+     * @chainable
+     */
+    reset() {
+        this._legCollection = [];
+
+        return this;
+    }
+
+    // ------------------------------ PUBLIC ------------------------------
+
+    // FIXME: COMPLETE THIS METHOD
+    /**
+     * Merge the provided route model into this route model, if possible
+     *
+     * @for RouteModel
+     * @method absorbRouteModel
+     * @param routeModel {RouteModel}
+     */
+    absorbRouteModel(/* routeModel */) {
```

Comment:
this, once implemented, should likely return a _new instance_ of a route model.  that should make things much easier, logic-wise because all you need to do is figure out how the route strings should be combined.

Reply:
For _proper_ route merging, it's actually far more complicated than merging the route strings themselves. For example, leaving the route in the middle of an airway leg: `instead of exiting the airway at this point, exit early here and join this other STAR`, etc.

As this has now actually been implemented, it does not return a new instance, but rather does just what it says: it takes in the supplied `RouteModel` instance, examines it and `this` in order to see if they have any point(s) of continuity, and if so, it adjusts the original routes legs as necessary and inserts the provided legs at the appropriate places. So it takes the provided route, breaks it down, and adds it to `this` route model.